go/libkb/chain_link.go

```diff
@@ -389,42 +427,130 @@ func (c *ChainLink) UnpackComputedKeyInfos(jw *jsonw.Wrapper) (err error) {
 	return
 }
 
-func (c *ChainLink) Unpack(trusted bool, selfUID keybase1.UID) (err error) {
-	tmp := ChainLinkUnpacked{}
-
-	c.packed.AtKey("sig").GetStringVoid(&tmp.sig, &err)
-	tmp.sigID, err = GetSigID(c.packed.AtKey("sig_id"), true)
-	c.packed.AtKey("payload_json").GetStringVoid(&tmp.payloadJSONStr, &err)
+type chainLinkPacked struct {
+	SigID         keybase1.SigID `json:"sig_id"`
+	Sig           string         `json:"sig"`
+	SigVersion    int            `json:"sigVersion"`
+	PayloadJSON   string         `json:"payload_json"`
+	ProofTextFull string         `json:"proof_text_full"`
+	SigVerified   bool           `json:"sig_verified"`
+}
 
+func (c *ChainLink) unpackStubbed(raw string, err error) error {
 	if err != nil {
 		return err
 	}
-
-	c.payloadJSON, err = jsonw.Unmarshal([]byte(tmp.payloadJSONStr))
+	var ol *OuterLinkV2WithMetadata
+	ol, err = DecodeStubbedOuterLinkV2(raw)
 	if err != nil {
 		return err
 	}
+	c.id = ol.Curr
+	c.unpacked = &ChainLinkUnpacked{
+		prev:        ol.Prev,
+		seqno:       ol.Seqno,
+		sigVersion:  ol.Version,
+		outerLinkV2: ol,
+		stubbed:     true,
+	}
+	return nil
+}
 
-	err = c.UnpackPayloadJSON(&tmp)
-	if err != nil {
-		return err
+func (c *ChainLink) Unpack(trusted bool, selfUID keybase1.UID) (err error) {
+	if jw := c.packed.AtKey("s2"); !jw.IsNil() {
+		return c.unpackStubbed(jw.GetString())
+	}
+
+	tmp := ChainLinkUnpacked{}
+	tmp.sigID, err = GetSigID(c.packed.AtKey("sig_id"), true)
+	c.packed.AtKey("sig").GetStringVoid(&tmp.sig, &err)
+
+	sigVersion := 1
+	if jw := c.packed.AtKey("sig_version"); !jw.IsNil() {
+		sigVersion, err = jw.GetInt()
+		if err != nil {
+			return err
+		}
 	}
+	tmp.sigVersion = sigVersion
 
-	// Set the unpacked.kid member if it's not already set. If it is set, check
-	// that the value is consistent with what's in the outer JSON blob.
-	serverKID, err := GetKID(c.packed.AtKey("kid"))
 	if err != nil {
 		return err
 	}
-	if tmp.kid.IsNil() {
-		tmp.kid = serverKID
-	} else if tmp.kid != serverKID {
+	if jw := c.packed.AtKey("payload_json"); !jw.IsNil() {
+		payloadJSONStr, err := jw.GetString()
+		if err != nil {
+			return err
+		}
+		payloadJSON, err := jsonw.Unmarshal([]byte(payloadJSONStr))
+		if err != nil {
+			return err
+		}
+		if err := tmp.unpackPayloadJSON(payloadJSON, payloadJSONStr); err != nil {
+			return err
+		}
+		c.payloadJSON = payloadJSON
+	}
+
+	var sigKID, serverKID, payloadKID keybase1.KID
+
+	if tmp.sigVersion == 2 {
+		var ol2 *OuterLinkV2WithMetadata
+		ol2, err = DecodeOuterLinkV2(tmp.sig)
+		if err != nil {
+			return err
+		}
+		linkID := ol2.Curr
+		if c.id != nil && !c.id.Eq(linkID) {
+			return SigchainV2MismatchedHashError{}
```

Comment:
If you make a bit change to the inner chain link, it fails here (there's a test for it).